static/_posts/2024-04-01-vgaterm-intro-001.md

```diff
@@ -0,0 +1,171 @@
+
+:----:|:-----:
+![Timing Board front](/assets/images/vgaterm-timing-front.jpg) | ![Timing Board back](/assets/images/vgaterm-timing-back.jpg)
+<span style="color: {{ site.caption_color }};">Vgaterm Timing circuit top</span> | <span style="color: {{ site.caption_color }};">Vgaterm Timing cicuit bottom</span>
+
+VGA timing is controlled by the "timing board" circuit. This establishes the `H_SYNC` and `V_SYNC` signals that VGA requires, the `START` signal which alerts the CPU to begin  emitting pixels to the FIFO, and the `visible` flag, which is used to compute `CLK_VIS` to read pixel bytes out of the FIFO.
+
+### Pixels Out
+Every frame (60 times a second) the esp32c3 microcontroller needs to output 256,000 bytes (640x400 pixels, 1 byte per pixel) to hardware so they may be displayed. We decided that this is the most important software role as too much jitter in timing will cause artifacts and wobbling in the piture.
+
+To cross the clock domains, from the CPU SPI to VGA pixel clock, we used a FIFO memory chip. This lets us write bytes from the CPU while VGA hardware reads bytes exactly when it needs them.
+
+{% include figure.html image="/assets/images/fifo-chip-highlight.jpg" width=450 position="right" caption="IDT 7205 FIFO chip" %}
+
+This worked well, but there are a few constraints we had to consider when designing a circuit around the FIFO:
+* The FIFO read and write setup and hold times had to be obeyed by `CLK_VIS` and the SPI clock signals respectively
+* The SPI only supported clock speeds in power-of-two divisions of 80 MHz (80, 40, 20, 10, etc)
+* If we emitted pixels too quickly, and VGA hardware could not read from the FIFO fast enough, we would blow the queue and drop pixels if the CPU didn't pause writing periodically.
+* If we emitted pixels too slowly, and VGA hardware consumed pixels faster than was being written, then we'd have to start loading pixels before VGA started consuming them.
+* The size of the FIFO effects how much wiggle room we get in the above two cases, especially as the FIFO is too expensive to have enough room for the whole frame.
```

Comment:
Teknically yes, but the more common use of `effect` is as a noun; consider:

 ```suggestion
* The size of the FIFO affects how much wiggle room we get in the above two cases, especially as the FIFO is too expensive to have enough room for the whole frame.
```

Though this is a little redundant with the description I suggested earlier; since the FIFO was such a primary constraint I think it works (maybe with a callback?), but you could consider using foreshadowing above instead.

Reply:
or even

```suggestion
* The FIFO is too expensive to have enough room for the whole frame, requiring more coordination between the readers & writers 
```